scripts/validate_docs_snippet_line_numbers.py

```diff
@@ -1,205 +1,189 @@
+def _determine_result(ref: Reference) -> Result:
+    if ref.target_lines is None:
+        return Result(
+            ref=ref,
+            status=Status.COMPLETE,
         )
-        return False
-
-    return _validate_opening_tag(start_line, file_contents) and _validate_closing_tag(
-        end_line, file_contents
-    )
-
 
-def _validate_opening_tag(start_line: int, file_contents: List[str]) -> bool:
-    ptr: int = start_line - 1
+    with open(ref.target_path) as f:
+        lines = f.readlines()
+
+    start, end = ref.target_lines
+
+    try:
+        open_tag = lines[start - 2]
+        valid_open = "<snippet" in open_tag
+    except IndexError:
+        valid_open = False
+
+    try:
+        close_tag = lines[end]
+        valid_close = "snippet>" in close_tag
+    except IndexError:
+        valid_close = False
+
+    status: Status
+    if valid_open and valid_close:
+        status = Status.COMPLETE
+    elif valid_open:
+        status = Status.MISSING_CLOSING
+    elif valid_close:
+        status = Status.MISSING_OPENING
+    else:
+        status = Status.MISSING_BOTH
 
-    line: str
-    while ptr >= 0:
-        line = file_contents[ptr - 1]
-        if OPENING_TAG in line:
-            return True
-        elif not line.isspace():
-            return False
-        ptr -= 1
+    return Result(
+        ref=ref,
+        status=status,
+    )
 
-    return False
 
+def evaluate_results(results: List[Result]) -> None:
+    summary = {}
 
-def _validate_closing_tag(end_line: int, file_contents: List[str]) -> bool:
-    ptr: int = end_line + 1
+    for res in results:
+        key = res.status.name
+        val = res.to_dict()
+        if key not in summary:
+            summary[key] = []
+        summary[key].append(val)
 
-    line: str
-    while ptr < len(file_contents):
-        line = file_contents[ptr - 1]
-        if CLOSING_TAG in line:
-            return True
-        elif not line.isspace():
-            return False
-        ptr += 1
+    pprint.pprint(summary)
 
-    return False
+    print("\n[SUMMARY]")
+    for key, val in summary.items():
+        print(f"* {key}: {len(val)}")
 
+    violations = len(results) - len(summary[Status.COMPLETE.name])
+    assert (
+        violations <= VIOLATION_THRESHOLD
+    ), f"Expected {VIOLATION_THRESHOLD} or fewer snippet violations, got {violations}"
 
-def print_diagnostic_report(broken_refs: List[DocusaurusRef]) -> None:
-    """Prints results to console for the user to review.
+    # There should only be COMPLETE (valid snippets) or MISSING_BOTH (snippets that haven't recieved surrounding tags yet)
+    # The presence of MISSING_OPENING or MISSING_CLOSING signifies a misaligned line number reference
+    assert (
+        summary.get(Status.MISSING_OPENING.name, 0) == 0
+    ), "Found a snippet without an opening snippet tag"
+    assert (
+        summary.get(Status.MISSING_CLOSING.name, 0) == 0
+    ), "Found a snippet without an closing snippet tag"
```

Comment:
Before all the docs changes in this PR, we have this opening and closing tag violations - we should have 0 since we either have proper snippets or referenced code that hasn't yet gotten a chance to receive surrounding `<snippet>` tags